File: utils/html_report_generator.py
#!/usr/bin/env python3
"""
HTML Report Generator Utility
Generates styled HTML reports from pandas DataFrames for GDC automation testing
"""

import pandas as pd
from datetime import datetime
import os

class HTMLReportGenerator:
    """Utility class for generating HTML reports from test data"""
    
    def __init__(self):
        """Initialize the HTML report generator"""
        self.status_colors = {
            "present_in_both": "#d4edda",
            "opensearch_only": "#cce5ff", 
            "legacy_only": "#fff3cd"
        }
        
        self.schema_colors = {
            "watch": "#fff3cd",
            "pep": "#d1ecf1", 
            "sanction": "#f8d7da",
            "icij": "#d4edda",
            "mex": "#e2e3e5",
            "rights": "#fce4ec",
            "soe": "#f3e5f5",
            "col": "#fff8e1",
            "media": "#e0f2f1"
        }
    
    def generate_unified_comparison_report(self, df, filename):
        """
        Generate HTML report from DataFrame for unified comparison
        """
        try:
            html_content = f"""
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Unified OpenSearch vs Legacy GDC Comparison Report</title>
    <style>
        body {{ font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif; margin: 0; padding: 20px; background-color: #f5f5f5; }}
        .container {{ max-width: 1400px; margin: 0 auto; background-color: white; border-radius: 8px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); overflow: hidden; }}
        .header {{ background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 30px; text-align: center; }}
        .header h1 {{ margin: 0; font-size: 2.5em; font-weight: 300; }}
        .header p {{ margin: 10px 0 0 0; font-size: 1.2em; opacity: 0.9; }}
        .summary {{ padding: 20px 30px; background-color: #f8f9fa; border-bottom: 1px solid #dee2e6; }}
        .summary h2 {{ margin: 0 0 15px 0; color: #495057; font-size: 1.5em; }}
        .stats {{ display: grid; grid-template-columns: repeat(auto-fit, minmax(200px, 1fr)); gap: 15px; }}
        .stat-card {{ background: white; padding: 15px; border-radius: 6px; text-align: center; box-shadow: 0 1px 3px rgba(0,0,0,0.1); }}
        .stat-number {{ font-size: 2em; font-weight: bold; color: #667eea; }}
        .stat-label {{ color: #6c757d; font-size: 0.9em; margin-top: 5px; }}
        .table-container {{ padding: 30px; overflow-x: auto; }}
        table {{ width: 100%; border-collapse: collapse; font-size: 0.9em; }}
        th {{ background-color: #495057; color: white; padding: 12px 8px; text-align: left; font-weight: 600; position: sticky; top: 0; z-index: 10; }}
        td {{ padding: 10px 8px; border-bottom: 1px solid #dee2e6; vertical-align: top; }}
        tr:nth-child(even) {{ background-color: #f8f9fa; }}
        tr:hover {{ background-color: #e3f2fd; }}
        .type-person {{ background-color: #e8f5e8; color: #2e7d32; padding: 2px 6px; border-radius: 4px; font-size: 0.8em; font-weight: 500; }}
        .type-entity {{ background-color: #e3f2fd; color: #1565c0; padding: 2px 6px; border-radius: 4px; font-size: 0.8em; font-weight: 500; }}
        .schema-badge {{ display: inline-block; padding: 2px 6px; border-radius: 4px; font-size: 0.8em; font-weight: 500; margin: 1px; }}
        .schema-watch {{ background-color: #fff3cd; color: #856404; }}
        .schema-pep {{ background-color: #d1ecf1; color: #0c5460; }}
        .schema-sanction {{ background-color: #f8d7da; color: #721c24; }}
        .schema-icij {{ background-color: #d4edda; color: #155724; }}
        .schema-mex {{ background-color: #e2e3e5; color: #383d41; }}
        .schema-rights {{ background-color: #fce4ec; color: #c2185b; }}
        .schema-soe {{ background-color: #f3e5f5; color: #7b1fa2; }}
        .schema-col {{ background-color: #fff8e1; color: #f57f17; }}
        .schema-media {{ background-color: #e0f2f1; color: #00695c; }}
        .empty-cell {{ color: #6c757d; font-style: italic; }}
        .status-badge {{ display: inline-block; padding: 4px 8px; border-radius: 4px; font-size: 0.8em; font-weight: 600; text-align: center; min-width: 80px; }}
        .status-both {{ background-color: #d4edda; color: #155724; border: 1px solid #c3e6cb; }}
        .status-opensearch-only {{ background-color: #cce5ff; color: #004085; border: 1px solid #b3d7ff; }}
        .status-legacy-only {{ background-color: #fff3cd; color: #856404; border: 1px solid #ffeaa7; }}
        .search-term-section {{ margin-bottom: 40px; }}
        .search-term-header {{ background: #f8f9fa; padding: 20px; border-radius: 8px 8px 0 0; border-bottom: 2px solid #dee2e6; }}
        .search-term-header h3 {{ margin: 0 0 10px 0; color: #495057; font-size: 1.5em; font-weight: 600; }}
        .term-summary {{ display: flex; flex-wrap: wrap; gap: 15px; }}
        .summary-item {{ background: white; padding: 8px 12px; border-radius: 4px; font-size: 0.9em; font-weight: 500; color: #495057; border: 1px solid #dee2e6; }}
        .search-term-table {{ width: 100%; border-collapse: collapse; font-size: 0.9em; }}
        .search-term-table th {{ background-color: #495057; color: white; padding: 12px 8px; text-align: left; font-weight: 600; }}
        .search-term-table td {{ padding: 10px 8px; border-bottom: 1px solid #dee2e6; vertical-align: top; }}
        .search-term-table tr:nth-child(even) {{ background-color: #f8f9fa; }}
        .search-term-table tr:hover {{ background-color: #e3f2fd; }}
        .footer {{ padding: 20px 30px; background-color: #f8f9fa; text-align: center; color: #6c757d; border-top: 1px solid #dee2e6; }}
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Unified OpenSearch vs Legacy GDC Comparison</h1>
            <p>Comprehensive Analysis Report - Generated on {datetime.now().strftime('%B %d, %Y at %I:%M %p')}</p>
        </div>
        
        <div class="table-container">
            <h2>Search Results by Term</h2>
"""
            
            # Generate separate tables for each search term
            for search_term in df['Search Term'].unique():
                term_df = df[df['Search Term'] == search_term].copy()
                
                # Calculate statistics for this term
                total_records = len(term_df)
                opensearch_records = len(term_df[term_df['OpenSearch ID'] != ''])
                legacy_records = len(term_df[term_df['Legacy ID'] != ''])
                matched_records = len(term_df[(term_df['OpenSearch ID'] != '') & (term_df['Legacy ID'] != '')])
                opensearch_only = len(term_df[(term_df['OpenSearch ID'] != '') & (term_df['Legacy ID'] == '')])
                legacy_only = len(term_df[(term_df['OpenSearch ID'] == '') & (term_df['Legacy ID'] != '')])
                
                html_content += f"""
            <div class="search-term-section">
                <div class="search-term-header">
                    <h3>{search_term}</h3>
                    <div class="term-summary">
                        <span class="summary-item">Total: {total_records}</span>
                        <span class="summary-item">OpenSearch: {opensearch_records}</span>
                        <span class="summary-item">Legacy: {legacy_records}</span>
                        <span class="summary-item">Matched: {matched_records}</span>
                        <span class="summary-item">OpenSearch Only: {opensearch_only}</span>
                        <span class="summary-item">Legacy Only: {legacy_only}</span>
                    </div>
                </div>
                <table class="search-term-table">
                    <thead>
                        <tr>
                            <th>#</th>
                            <th>Type</th>
                            <th>Status</th>
                            <th>OpenSearch Name</th>
                            <th>OpenSearch ID</th>
                            <th>OpenSearch Schema</th>
                            <th>Legacy Name</th>
                            <th>Legacy ID</th>
                            <th>Legacy Schema</th>
                        </tr>
                    </thead>
                    <tbody>
"""
                
                # Add rows for this search term
                row_number = 1
                for _, row in term_df.iterrows():
                    type_class = "type-person" if row['Type'] == 'Person' else "type-entity"
                    opensearch_schema_class = f"schema-{row['OpenSearch Schema'].lower()}" if row['OpenSearch Schema'] else ""
                    legacy_schema_class = f"schema-{row['Legacy Schema'].lower()}" if row['Legacy Schema'] else ""
                    
                    # Convert to string and check for "Not Present" text
                    opensearch_name = str(row['OpenSearch Name']) if row['OpenSearch Name'] else ''
                    opensearch_id = str(row['OpenSearch ID']) if row['OpenSearch ID'] else ''
                    opensearch_schema = str(row['OpenSearch Schema']) if row['OpenSearch Schema'] else ''
                    legacy_name = str(row['Legacy Name']) if row['Legacy Name'] else ''
                    legacy_id = str(row['Legacy ID']) if row['Legacy ID'] else ''
                    legacy_schema = str(row['Legacy Schema']) if row['Legacy Schema'] else ''
                    
                    # Check if values indicate presence (not empty and not "Not Present")
                    has_opensearch = opensearch_id and 'Not Present' not in opensearch_id
                    has_legacy = legacy_id and 'Not Present' not in legacy_id
                    
                    if has_opensearch and has_legacy:
                        status = "Present in Both"
                        status_class = "status-both"
                    elif has_opensearch and not has_legacy:
                        status = "OpenSearch Only"
                        status_class = "status-opensearch-only"
                    elif not has_opensearch and has_legacy:
                        status = "Legacy Only"
                        status_class = "status-legacy-only"
                    else:
                        status = "Not Present"
                        status_class = "status-legacy-only"
                    
                    # Keep existing "Not Present" text or set to "Not Present" if empty
                    opensearch_name = opensearch_name if opensearch_name else 'Not Present'
                    opensearch_id = opensearch_id if opensearch_id else 'Not Present'
                    opensearch_schema = opensearch_schema if opensearch_schema else 'Not Present'
                    legacy_name = legacy_name if legacy_name else 'Not Present'
                    legacy_id = legacy_id if legacy_id else 'Not Present'
                    legacy_schema = legacy_schema if legacy_schema else 'Not Present'
                    
                    html_content += f"""
                        <tr>
                            <td><strong>{row_number}</strong></td>
                            <td><span class="{type_class}">{row['Type']}</span></td>
                            <td><span class="status-badge {status_class}">{status}</span></td>
                            <td>{opensearch_name}</td>
                            <td>{opensearch_id}</td>
                            <td><span class="schema-badge {opensearch_schema_class}">{opensearch_schema}</span></td>
                            <td>{legacy_name}</td>
                            <td>{legacy_id}</td>
                            <td><span class="schema-badge {legacy_schema_class}">{legacy_schema}</span></td>
                        </tr>
"""
                    row_number += 1
                
                html_content += """
                    </tbody>
                </table>
            </div>
"""
            
            html_content += """
        </div>
        
        <div class="footer">
            <p>Generated by GDC Automation Excel-Driven Regression Testing Framework</p>
            <p>This report compares OpenSearch API results with Legacy GDC system data</p>
        </div>
    </div>
</body>
</html>
"""
            
            with open(filename, 'w', encoding='utf-8') as f:
                f.write(html_content)
                
            print(f"✅ HTML report generated: {filename}")
                
        except Exception as e:
            print(f"❌ Error generating HTML report: {e}")
            raise
    
    def generate_entity_report(self, entity_name, entity_type, comparison_data, filename):
        """
        Generate HTML report for individual entity comparison
        """
        try:
            html_content = f"""
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Entity Comparison Report - {entity_name}</title>
    <style>
        body {{ font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif; margin: 0; padding: 20px; background-color: #f5f5f5; }}
        .container {{ max-width: 1200px; margin: 0 auto; background-color: white; border-radius: 8px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); }}
        .header {{ background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 30px; text-align: center; }}
        .header h1 {{ margin: 0; font-size: 2em; }}
        .content {{ padding: 30px; }}
        .section {{ margin-bottom: 30px; }}
        .section h2 {{ color: #495057; border-bottom: 2px solid #dee2e6; padding-bottom: 10px; }}
        .stats {{ display: grid; grid-template-columns: repeat(auto-fit, minmax(150px, 1fr)); gap: 15px; margin: 20px 0; }}
        .stat-card {{ background: #f8f9fa; padding: 15px; border-radius: 6px; text-align: center; }}
        .stat-number {{ font-size: 1.5em; font-weight: bold; color: #667eea; }}
        .stat-label {{ color: #6c757d; font-size: 0.9em; }}
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>Entity Comparison Report</h1>
            <p>{entity_name} ({entity_type})</p>
        </div>
        <div class="content">
            <div class="section">
                <h2>Summary</h2>
                <div class="stats">
                    <div class="stat-card">
                        <div class="stat-number">{comparison_data.get('common_records', 0)}</div>
                        <div class="stat-label">Common Records</div>
                    </div>
                    <div class="stat-card">
                        <div class="stat-number">{comparison_data.get('new_in_opensearch', 0)}</div>
                        <div class="stat-label">New in OpenSearch</div>
                    </div>
                    <div class="stat-card">
                        <div class="stat-number">{comparison_data.get('missing_from_opensearch', 0)}</div>
                        <div class="stat-label">Missing from OpenSearch</div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</body>
</html>
"""
            
            with open(filename, 'w', encoding='utf-8') as f:
                f.write(html_content)
                
        except Exception as e:
            print(f"Error generating entity HTML report: {e}")
            raise

File: k6-tests/utils/html-report-generator.js
#!/usr/bin/env node

/**
 * K6 HTML Report Generator
 * Converts K6 JSON output to beautiful HTML reports
 */

const fs = require('fs');
const path = require('path');

class K6HTMLReportGenerator {
    constructor() {
        this.metrics = {};
        this.summary = {};
    }

    parseJSONFile(filePath) {
        try {
            const data = fs.readFileSync(filePath, 'utf8');
            const lines = data.trim().split('\n');
            
            for (const line of lines) {
                if (line.trim()) {
                    const parsed = JSON.parse(line);
                    this.processMetric(parsed);
                }
            }
        } catch (error) {
            console.error(`Error parsing JSON file: ${error.message}`);
            process.exit(1);
        }
    }

    processMetric(metric) {
        if (metric.type === 'Metric') {
            this.metrics[metric.metric] = metric.data;
        } else if (metric.type === 'Point') {
            if (!this.metrics[metric.metric]) {
                this.metrics[metric.metric] = { values: [] };
            }
            if (!this.metrics[metric.metric].values) {
                this.metrics[metric.metric].values = [];
            }
            // Prevent infinite recursion by limiting array size
            if (this.metrics[metric.metric].values.length < 10000) {
                this.metrics[metric.metric].values.push(metric.data);
            }
        }
    }

    calculateSummary() {
        const summary = {};
        
        // Calculate summary statistics
        for (const [metricName, metricData] of Object.entries(this.metrics)) {
            if (metricData && metricData.values && Array.isArray(metricData.values) && metricData.values.length > 0) {
                const values = metricData.values.map(v => v.value).filter(v => typeof v === 'number' && !isNaN(v));
                if (values.length > 0) {
                    try {
                        summary[metricName] = {
                            count: values.length,
                            min: Math.min(...values),
                            max: Math.max(...values),
                            avg: values.reduce((a, b) => a + b, 0) / values.length,
                            p90: this.percentile(values, 90),
                            p95: this.percentile(values, 95),
                            p99: this.percentile(values, 99)
                        };
                    } catch (error) {
                        console.warn(`Error calculating summary for ${metricName}:`, error.message);
                        summary[metricName] = {
                            count: values.length,
                            min: 0,
                            max: 0,
                            avg: 0,
                            p90: 0,
                            p95: 0,
                            p99: 0
                        };
                    }
                }
            }
        }
        
        this.summary = summary;
    }

    percentile(arr, p) {
        const sorted = arr.slice().sort((a, b) => a - b);
        const index = Math.ceil((p / 100) * sorted.length) - 1;
        return sorted[index];
    }

    generateHTML() {
        const timestamp = new Date().toLocaleString();
        
        return `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>K6 Performance Test Report</title>
    <style>
        * {
            margin: 0;
            padding: 0;
            box-sizing: border-box;
        }
        
        body {
            font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            min-height: 100vh;
            padding: 20px;
        }
        
        .container {
            max-width: 1200px;
            margin: 0 auto;
            background: white;
            border-radius: 12px;
            box-shadow: 0 10px 30px rgba(0,0,0,0.2);
            overflow: hidden;
        }
        
        .header {
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
            padding: 30px;
            text-align: center;
        }
        
        .header h1 {
            font-size: 2.5em;
            margin-bottom: 10px;
            font-weight: 300;
        }
        
        .header p {
            font-size: 1.2em;
            opacity: 0.9;
        }
        
        .content {
            padding: 30px;
        }
        
        .summary-grid {
            display: grid;
            grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
            gap: 20px;
            margin-bottom: 30px;
        }
        
        .metric-card {
            background: #f8f9fa;
            border-radius: 8px;
            padding: 20px;
            border-left: 4px solid #667eea;
        }
        
        .metric-card h3 {
            color: #495057;
            margin-bottom: 15px;
            font-size: 1.2em;
        }
        
        .metric-stats {
            display: grid;
            grid-template-columns: repeat(2, 1fr);
            gap: 10px;
        }
        
        .stat {
            display: flex;
            justify-content: space-between;
            padding: 8px 0;
            border-bottom: 1px solid #e9ecef;
        }
        
        .stat:last-child {
            border-bottom: none;
        }
        
        .stat-label {
            color: #6c757d;
            font-weight: 500;
        }
        
        .stat-value {
            color: #495057;
            font-weight: 600;
        }
        
        .charts-section {
            margin-top: 30px;
        }
        
        .charts-section h2 {
            color: #495057;
            margin-bottom: 20px;
            font-size: 1.5em;
        }
        
        .chart-container {
            background: #f8f9fa;
            border-radius: 8px;
            padding: 20px;
            margin-bottom: 20px;
        }
        
        .chart-title {
            color: #495057;
            margin-bottom: 15px;
            font-size: 1.1em;
        }
        
        .footer {
            background: #f8f9fa;
            padding: 20px;
            text-align: center;
            color: #6c757d;
            border-top: 1px solid #e9ecef;
        }
        
        .status-good { color: #28a745; }
        .status-warning { color: #ffc107; }
        .status-danger { color: #dc3545; }
        
        .progress-bar {
            width: 100%;
            height: 8px;
            background: #e9ecef;
            border-radius: 4px;
            overflow: hidden;
            margin: 5px 0;
        }
        
        .progress-fill {
            height: 100%;
            background: linear-gradient(90deg, #28a745, #20c997);
            transition: width 0.3s ease;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>🚀 K6 Performance Test Report</h1>
            <p>Generated on ${timestamp}</p>
        </div>
        
        <div class="content">
            <div class="summary-grid">
                ${this.generateMetricCards()}
            </div>
            
            <div class="charts-section">
                <h2>📊 Performance Metrics</h2>
                ${this.generateCharts()}
            </div>
        </div>
        
        <div class="footer">
            <p>Generated by K6 Performance Testing Framework</p>
            <p>This report shows detailed performance metrics from your load test</p>
        </div>
    </div>
</body>
</html>`;
    }

    generateMetricCards() {
        const cards = [];
        
        // Key metrics to display
        const keyMetrics = [
            'http_reqs',
            'http_req_duration',
            'http_req_failed',
            'vus',
            'iterations'
        ];
        
        for (const metric of keyMetrics) {
            if (this.summary[metric]) {
                const data = this.summary[metric];
                cards.push(this.generateMetricCard(metric, data));
            }
        }
        
        return cards.join('');
    }

    generateMetricCard(metricName, data) {
        const title = this.getMetricTitle(metricName);
        const icon = this.getMetricIcon(metricName);
        
        return `
        <div class="metric-card">
            <h3>${icon} ${title}</h3>
            <div class="metric-stats">
                <div class="stat">
                    <span class="stat-label">Count</span>
                    <span class="stat-value">${data.count.toLocaleString()}</span>
                </div>
                <div class="stat">
                    <span class="stat-label">Average</span>
                    <span class="stat-value">${this.formatValue(metricName, data.avg)}</span>
                </div>
                <div class="stat">
                    <span class="stat-label">Min</span>
                    <span class="stat-value">${this.formatValue(metricName, data.min)}</span>
                </div>
                <div class="stat">
                    <span class="stat-label">Max</span>
                    <span class="stat-value">${this.formatValue(metricName, data.max)}</span>
                </div>
                <div class="stat">
                    <span class="stat-label">95th Percentile</span>
                    <span class="stat-value">${this.formatValue(metricName, data.p95)}</span>
                </div>
            </div>
        </div>`;
    }

    getMetricTitle(metricName) {
        const titles = {
            'http_reqs': 'HTTP Requests',
            'http_req_duration': 'Response Time',
            'http_req_failed': 'Failed Requests',
            'vus': 'Virtual Users',
            'iterations': 'Iterations'
        };
        return titles[metricName] || metricName;
    }

    getMetricIcon(metricName) {
        const icons = {
            'http_reqs': '🌐',
            'http_req_duration': '⏱️',
            'http_req_failed': '❌',
            'vus': '👥',
            'iterations': '🔄'
        };
        return icons[metricName] || '📊';
    }

    formatValue(metricName, value) {
        if (metricName === 'http_req_duration') {
            return `${value.toFixed(2)}ms`;
        } else if (metricName === 'http_req_failed') {
            return `${(value * 100).toFixed(2)}%`;
        } else {
            return value.toLocaleString();
        }
    }

    generateCharts() {
        return `
        <div class="chart-container">
            <div class="chart-title">📈 Performance Overview</div>
            <p>This report provides detailed insights into your application's performance under load.</p>
            <p><strong>Key Insights:</strong></p>
            <ul>
                <li>Total HTTP Requests: ${this.summary.http_reqs?.count || 'N/A'}</li>
                <li>Average Response Time: ${this.formatValue('http_req_duration', this.summary.http_req_duration?.avg || 0)}</li>
                <li>Failed Requests: ${this.formatValue('http_req_failed', this.summary.http_req_failed?.avg || 0)}</li>
                <li>Peak Virtual Users: ${this.summary.vus?.max || 'N/A'}</li>
            </ul>
        </div>`;
    }

    generateReport(jsonFilePath, outputPath) {
        console.log('📊 Parsing K6 JSON data...');
        this.parseJSONFile(jsonFilePath);
        
        console.log('📈 Calculating summary statistics...');
        this.calculateSummary();
        
        console.log('🎨 Generating HTML report...');
        const html = this.generateHTML();
        
        fs.writeFileSync(outputPath, html);
        console.log(`✅ HTML report generated: ${outputPath}`);
    }
}

// Command line usage
if (require.main === module) {
    const args = process.argv.slice(2);
    
    if (args.length < 2) {
        console.log('Usage: node html-report-generator.js <json-file> <output-html-file>');
        console.log('Example: node html-report-generator.js results/k6-test.json results/k6-report.html');
        process.exit(1);
    }
    
    const [jsonFile, htmlFile] = args;
    const generator = new K6HTMLReportGenerator();
    generator.generateReport(jsonFile, htmlFile);
}

module.exports = K6HTMLReportGenerator;
